File: java/src/main/java/com/example/vulnapp/vulns/CommandInjectionController.java
package com.example.vulnapp.vulns;

import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.TimeUnit;
import java.util.regex.Pattern;

/**
 * OS Command Injection — OWASP 2021 A03 Injection | OWASP 2025 A05 Injection.
 *
 * Mirrors the comment/annotation format of {@link SqlInjectionController}: each
 * permutation carries a standard header describing the flaw, its CWE, an
 * exploitation-likelihood rating, greenfield-vs-legacy prevalence, an example
 * payload, and the corresponding safe pattern. Every dangerous sink is a real,
 * exploitable {@code ProcessBuilder} that reaches the OS — nothing is stubbed.
 *
 * PRIMARY CWE: CWE-78 Improper Neutralization of Special Elements used in an OS
 * Command ('OS Command Injection').
 */
@RestController
@RequestMapping("/injection/command")
public class CommandInjectionController {

    // ========================================================================
    // PERMUTATION 1 — Shell command built by string concatenation (ping)
    // OWASP 2021 A03 Injection  ->  OWASP 2025 A05 Injection
    // CWE-78: Improper Neutralization of Special Elements used in an OS Command
    // EXPLOITATION LIKELIHOOD: HIGH — pre-auth, deterministic one-liner; the
    //   untrusted host is glued into a string and handed to `bash -c`, so a
    //   `;`/`|`/`$()` metacharacter runs an arbitrary second command. The
    //   injected command executes even if `ping` itself is absent.
    // PREVALENCE TODAY: greenfield LOW (modern stacks rarely shell out; when they
    //   must they use argv exec + libraries, and SAST/linters flag `bash -c` with
    //   a concatenated string) | legacy/10yr tech-debt HIGH (network/ops/admin
    //   features wrapping shell one-liners with string concat are everywhere in
    //   long-lived codebases).
    // EXAMPLE:
    //   GET /injection/command/ping?host=127.0.0.1;id
    //   GET /injection/command/ping?host=$(cat%20/etc/passwd)
    // FIX: never build a shell string from input — validate against an allow-list
    //   and exec an argv array with no shell (see /injection/command/ping-safe).
    // ========================================================================
    @GetMapping("/ping")
    public Map<String, Object> ping(@RequestParam(defaultValue = "127.0.0.1") String host) {
        // VULNERABLE: untrusted host concatenated into a command line run through a shell.
        String cmd = "ping -c 1 " + host;
        Map<String, Object> out = new LinkedHashMap<>();
        out.put("command", cmd);
        out.put("shell", "bash -c");
        out.putAll(exec(new ProcessBuilder("bash", "-c", cmd)));
        return out;
    }

    // ========================================================================
    // PERMUTATION 2 — Second shell sink with a different tool (nslookup)
    // OWASP 2021 A03 Injection  ->  OWASP 2025 A05 Injection
    // CWE-78: Improper Neutralization of Special Elements used in an OS Command
    // EXPLOITATION LIKELIHOOD: HIGH — same class as P1 but via `sh -c`; a lookup/
    //   diagnostics box concatenates the domain, so `example.com;cat /etc/passwd`
    //   leaks arbitrary files. Pre-auth, deterministic, no tooling required.
    // PREVALENCE TODAY: greenfield LOW (DNS/resolution is done with library calls,
    //   e.g. InetAddress, not by shelling out) | legacy/10yr tech-debt HIGH
    //   (hand-rolled `nslookup`/`host`/`dig` wrappers persist in old admin panels).
    // EXAMPLE:
    //   GET /injection/command/dns?domain=example.com;cat%20/etc/passwd
    //   GET /injection/command/dns?domain=x|id
    // FIX: resolve names with java.net.InetAddress; never pass input to a shell.
    // ========================================================================
    @GetMapping("/dns")
    public Map<String, Object> dns(@RequestParam(defaultValue = "example.com") String domain) {
        // VULNERABLE: untrusted domain concatenated into a command line run through a shell.
        String cmd = "nslookup " + domain;
        Map<String, Object> out = new LinkedHashMap<>();
        out.put("command", cmd);
        out.put("shell", "sh -c");
        out.putAll(exec(new ProcessBuilder("sh", "-c", cmd)));
        return out;
    }

    // ========================================================================
    // PERMUTATION 3 — Argument injection into argv exec, NO shell (tar)
    // OWASP 2021 A03 Injection  ->  OWASP 2025 A05 Injection
    // CWE-78: Improper Neutralization of Special Elements used in an OS Command
    //   (specifically CWE-88 Argument Injection, a CWE-78 variant)
    // EXPLOITATION LIKELIHOOD: MEDIUM — there is NO shell, so `;`/`|`/`$()` do
    //   nothing; the attacker must know that argv values beginning with `-` are
    //   parsed by `tar` as OPTIONS. Given that knowledge it is reliable: tar's
    //   own `--checkpoint-action=exec=` runs an arbitrary command. Reaching code
    //   execution takes more skill than a shell metacharacter, hence MEDIUM.
    // PREVALENCE TODAY: greenfield MEDIUM ("we use array exec, so we're safe" is a
    //   widespread myth; devs pass user input straight into a `tar`/`git`/`curl`/
    //   `ffmpeg` argv without a `--` end-of-options guard, and SAST rarely flags
    //   argv injection) | legacy/10yr tech-debt HIGH (countless CLI-wrapper
    //   backup/export features built by splitting a user string into args).
    // EXAMPLE (array-exec, still RCE — note: no metacharacters):
    //   GET /injection/command/backup?filename=--checkpoint=1 --checkpoint-action=exec=id pom.xml
    //   -> tar reaches a checkpoint and runs `id`; stdout is captured below.
    // FIX: insert a literal "--" end-of-options marker before user paths AND
    //   allow-list the values so they cannot be interpreted as flags.
    // ========================================================================
    @GetMapping("/backup")
    public Map<String, Object> backup(@RequestParam(defaultValue = "pom.xml") String filename) {
        List<String> argv = new ArrayList<>();
        argv.add("tar");
        argv.add("-czf");
        argv.add("/tmp/backup.tgz");
        // VULNERABLE: attacker-controlled filename split into argv tokens and passed to
        // tar with NO shell. Array-exec is NOT automatically safe: tokens starting with
        // '-' are consumed as tar OPTIONS (e.g. --checkpoint-action=exec=<cmd>), so tar
        // itself executes the command. There is no "--" end-of-options guard here.
        for (String tok : filename.trim().split("\\s+")) {
            if (!tok.isEmpty()) {
                argv.add(tok);
            }
        }
        Map<String, Object> out = new LinkedHashMap<>();
        out.put("note", "no shell used — plain ProcessBuilder(argv); array-exec is NOT injection-proof");
        out.putAll(exec(new ProcessBuilder(argv)));
        return out;
    }

    // ========================================================================
    // SAFE REFERENCE — strict allow-list + argv exec, no shell. Shown so the
    // vulnerable/safe pair can be diffed by SAST tooling and learners.
    //
    // Two independent controls make this safe:
    //   1) the host must match ^[A-Za-z0-9][A-Za-z0-9._-]{0,252}$ — it cannot
    //      contain shell metacharacters AND cannot begin with '-', so it can
    //      never be reinterpreted as a command-line option (defeats P3-style
    //      argument injection too);
    //   2) it is passed as a single element of an argv array to ProcessBuilder
    //      with no shell at all (defeats P1/P2-style metacharacter injection).
    // ========================================================================
    private static final Pattern SAFE_HOST = Pattern.compile("^[A-Za-z0-9][A-Za-z0-9._-]{0,252}$");

    @GetMapping("/ping-safe")
    public Map<String, Object> pingSafe(@RequestParam(defaultValue = "127.0.0.1") String host) {
        Map<String, Object> out = new LinkedHashMap<>();
        out.put("host", host);
        if (!SAFE_HOST.matcher(host).matches()) {
            // SAFE: reject anything that is not a bare hostname/IP before it reaches the OS.
            out.put("rejected", "host failed allow-list ^[A-Za-z0-9][A-Za-z0-9._-]{0,252}$");
            return out;
        }
        // SAFE: no shell; host is one argv element that already passed validation.
        out.putAll(exec(new ProcessBuilder("ping", "-c", "1", host)));
        return out;
    }

    /**
     * Runs a process, captures stdout/stderr/exit code, and returns them for the
     * response. {@code readAllBytes()} blocks until each stream hits EOF (i.e.
     * the child closes it on exit), so this also serves as the wait for small
     * command outputs; a bounded {@code waitFor} guards against a hang.
     */
    private Map<String, Object> exec(ProcessBuilder pb) {
        Map<String, Object> out = new LinkedHashMap<>();
        out.put("argv", new ArrayList<>(pb.command()));
        try {
            Process p = pb.start();
            String stdout = new String(p.getInputStream().readAllBytes(), StandardCharsets.UTF_8);
            String stderr = new String(p.getErrorStream().readAllBytes(), StandardCharsets.UTF_8);
            boolean finished = p.waitFor(20, TimeUnit.SECONDS);
            if (!finished) {
                p.destroyForcibly();
            }
            out.put("exitCode", finished ? p.exitValue() : null);
            out.put("stdout", stdout);
            out.put("stderr", stderr);
        } catch (Exception e) {
            out.put("error", e.getClass().getSimpleName() + ": " + e.getMessage());
        }
        return out;
    }
}

File: csharp/Controllers/CommandInjectionController.cs
using Microsoft.AspNetCore.Mvc;
using System.Diagnostics;
using System.Text.RegularExpressions;

namespace VulnApp.Controllers;

/// <summary>
/// Command Injection — OWASP 2021 A03 Injection | OWASP 2025 A05 Injection.
///
/// Mirrors the annotation format and attribute-routed action style of
/// SqlInjectionController. Every handler shells out to (or execs) a real OS
/// process so the injection is genuinely exploitable when the app runs.
///
/// NOTE ON THIS LAB IMAGE: ping/host/nslookup are frequently absent from minimal
/// containers, but that does NOT blunt the vulnerability — in the shell-based
/// permutations the *injected* command (after ';', '|', '$()', …) still runs via
/// /bin/sh, which is what turns these into RCE.
/// </summary>
[ApiController]
public class CommandInjectionController : ControllerBase
{
    // ========================================================================
    // PERMUTATION 1 — OS command built by concatenation, run through a SHELL
    // OWASP 2021 A03 Injection -> OWASP 2025 A05 Injection
    // CWE-78: Improper Neutralization of Special Elements used in an OS Command
    // EXPLOITATION LIKELIHOOD: CRITICAL — unauthenticated GET, zero tooling; the
    //   textbook  host=127.0.0.1;id  yields deterministic RCE as the app process.
    // PREVALENCE TODAY: greenfield LOW (idiomatic .NET uses ProcessStartInfo.
    //   ArgumentList; SAST flags building a shell string) | legacy/10yr tech-debt
    //   HIGH (old admin/ops code shells out to ping/ffmpeg/git via concatenation).
    // EXAMPLE: GET /injection/command/ping?host=127.0.0.1;id  (encode as 127.0.0.1%3Bid) -> runs `id`
    // FIX: no shell — fixed argv via ArgumentList + allow-list (see /injection/command/ping-safe).
    // ========================================================================
    [HttpGet("/injection/command/ping")]
    public IActionResult Ping(string host = "127.0.0.1")
    {
        // The classic "shell out to a CLI tool" pattern: build a command LINE by
        // concatenation and hand the whole string to /bin/sh -c.
        var command = $"ping -c 1 {host}"; // VULNERABLE: host concatenated into a /bin/sh command line
        var (stdout, stderr, exit) = RunShell(command);
        return new JsonResult(new
        {
            shell = "/bin/sh -c",
            command,
            exit,
            stdout,
            stderr,
            note = "everything after ; | && $() runs via /bin/sh — e.g. host=127.0.0.1;id executes id"
        });
    }

    // ========================================================================
    // PERMUTATION 2 — a SECOND shell command using a different tool (DNS lookup)
    // OWASP 2021 A03 Injection -> OWASP 2025 A05 Injection
    // CWE-78: Improper Neutralization of Special Elements used in an OS Command
    // EXPLOITATION LIKELIHOOD: CRITICAL — identical pre-auth /bin/sh sink; the
    //   payload  domain=example.com;cat /etc/passwd  exfiltrates arbitrary files
    //   and is trivially escalated to full RCE.
    // PREVALENCE TODAY: greenfield LOW (teams resolve names with a DNS library, or
    //   use fixed argv, not /bin/sh) | legacy/10yr tech-debt HIGH (network-tool
    //   wrappers assembled as shell strings are everywhere in old code).
    // EXAMPLE: GET /injection/command/dns?domain=example.com;cat /etc/passwd -> dumps /etc/passwd
    // FIX: resolve names with System.Net.Dns, or fixed-argv host/nslookup + strict validation — no shell.
    // ========================================================================
    [HttpGet("/injection/command/dns")]
    public IActionResult Dns(string domain = "example.com")
    {
        // A different tool ("host"), same fatal mistake: concatenate + /bin/sh -c.
        var command = $"host {domain}"; // VULNERABLE: domain concatenated into a /bin/sh command line
        var (stdout, stderr, exit) = RunShell(command);
        return new JsonResult(new
        {
            shell = "/bin/sh -c",
            command,
            exit,
            stdout,
            stderr,
            note = "domain=example.com;cat /etc/passwd reads arbitrary files through the shell"
        });
    }

    // ========================================================================
    // PERMUTATION 3 — ARGUMENT (option) injection into tar with NO shell (argv)
    // OWASP 2021 A03 Injection -> OWASP 2025 A05 Injection
    // CWE-78: Improper Neutralization of Special Elements used in an OS Command
    // EXPLOITATION LIKELIHOOD: HIGH — no shell metacharacters required; relies on
    //   the app splitting user input into argv + a known tar option gadget, but is
    //   deterministic RCE once those hold. Proves argv-exec is NOT automatically safe.
    // PREVALENCE TODAY: greenfield MEDIUM (argv-exec is the *recommended* fix, so
    //   option injection is overlooked and the `--` end-of-options guard is routinely
    //   forgotten) | legacy/10yr tech-debt MEDIUM (same blind spot, more CLI shelling).
    // EXAMPLE: GET /injection/command/backup?filename=--checkpoint=1 --checkpoint-action=exec=id
    //          (URL-encode the spaces) -> tar runs `id` with no /bin/sh involved.
    // FIX: insert a literal "--" before any user token so tar stops parsing options,
    //      and allow-list filenames (reject leading '-').
    // ========================================================================
    [HttpGet("/injection/command/backup")]
    public IActionResult Backup(string filename = "")
    {
        // "We build the argv ourselves and never touch /bin/sh, so we're safe."
        // A common misconception. This handler backs up the app's data directory
        // and lets the caller add extra space-separated files to the archive.
        var argv = new List<string> { "-czf", "/tmp/backup.tgz" };
        // VULNERABLE: user tokens become LEADING tar arguments; any token starting
        // with '-' is parsed as an OPTION (e.g. --checkpoint-action=exec=CMD, which
        // tar itself runs through a shell) rather than as a filename to archive.
        foreach (var tok in filename.Split(' ', StringSplitOptions.RemoveEmptyEntries))
            argv.Add(tok);
        argv.Add("data"); // real source, so tar writes records and reaches the checkpoint
        var (stdout, stderr, exit) = RunProcess("tar", argv);
        return new JsonResult(new
        {
            shell = "none (direct argv exec)",
            argv = new List<string> { "tar" }.Concat(argv),
            exit,
            stdout,
            stderr,
            note = "no shell was used, yet tar's --checkpoint-action=exec option still executes code"
        });
    }

    // ========================================================================
    // SAFE REFERENCE — strict allow-list + fixed argv + NO shell. Shown so the
    // vulnerable/safe pair can be diffed by SAST tooling and by learners.
    //   (1) allow-list validation rejects shell metachars and any leading '-';
    //   (2) no /bin/sh — the binary is exec'd directly with a fixed argv;
    //   (3) '--' end-of-options guard means a hostile value can never be an option.
    // ========================================================================
    [HttpGet("/injection/command/ping-safe")]
    public IActionResult PingSafe(string host = "127.0.0.1")
    {
        if (!Regex.IsMatch(host, "^(?!-)[A-Za-z0-9.-]{1,253}$"))
            return new JsonResult(new { safe = true, error = "invalid host — allow-list rejected it", host }) { StatusCode = 400 };

        var argv = new List<string> { "-c", "1", "--", host };
        var (stdout, stderr, exit) = RunProcess("ping", argv);
        return new JsonResult(new
        {
            safe = true,
            argv = new List<string> { "ping" }.Concat(argv),
            exit,
            stdout,
            stderr
        });
    }

    // --- helpers -------------------------------------------------------------

    // Run a command LINE through the system shell (the dangerous sink used by P1/P2).
    private static (string Stdout, string Stderr, int Exit) RunShell(string command)
        => RunProcess("/bin/sh", new List<string> { "-c", command });

    // Exec a program with an explicit argv (no shell) and capture stdout/stderr.
    private static (string Stdout, string Stderr, int Exit) RunProcess(string fileName, List<string> argv)
    {
        var psi = new ProcessStartInfo
        {
            FileName = fileName,
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            UseShellExecute = false,
            WorkingDirectory = Directory.GetCurrentDirectory(),
        };
        foreach (var a in argv) psi.ArgumentList.Add(a);

        try
        {
            using var p = Process.Start(psi);
            var outTask = p.StandardOutput.ReadToEndAsync();
            var errTask = p.StandardError.ReadToEndAsync();
            p.WaitForExit(10000);
            return (outTask.GetAwaiter().GetResult(), errTask.GetAwaiter().GetResult(),
                    p.HasExited ? p.ExitCode : -1);
        }
        catch (Exception e)
        {
            // e.g. the intended tool (ping/host) is not installed on a minimal image.
            return ("", e.Message, -1);
        }
    }
}
